### PR TITLE
Pv2 843 pv2 676 create multiple learners validator component

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CalculatePeriodStartAndEndDateTest.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace SFA.DAS.Payments.DataLocks.Domain.UnitTests.Services
+{
+    [TestFixture]
+    public class CalculatePeriodStartAndEndDateTest
+    {
+
+        [TestCase( 1, 1920, "2019/8/1","2019/8/31")]
+        [TestCase( 2, 1920, "2019/9/1", "2019/9/30")]
+        [TestCase(6, 1920, "2020/1/1","2020/1/31")]
+        [TestCase( 7, 1920, "2020/2/1", "2020/2/29")]
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CalculatePeriodStartAndEndDateTest.cs]: Worth adding 5, 11, 12

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/MultipleLearnersValidatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        }
+
+        [Test]
+        public void WhenNoDuplicateApprenticeshipsReturnNoDataLockErrors()
+        {
+            var earningPeriod = new EarningPeriod {Period = 1};
+
+            calculatePeriodStartAndEndDate
+                .Setup(x => x.GetPeriodDate(earningPeriod.Period, academicYear))
+                .Returns((DateTime.Parse("01 Aug 2018"), DateTime.Parse("31 Aug 2018")))
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/MultipleLearnersValidatorTest.cs]: Why not `new DateTime(2018, 8, 1)`?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipLevyPayerValidator.cs [hunk 3]
```diff
@@ -14,7 +14,7 @@ protected override List<ApprenticeshipPriceEpisodeModel> GetValidApprenticeshipP
         {
             return dataLockValidationModel.Apprenticeship.ApprenticeshipPriceEpisodes;
         }
-        protected override bool FailedValidation(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
+        protected override bool FailedValidationAsync(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipLevyPayerValidator.cs]: Why an async suffix without returning a `Task`?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipPausedValidator.cs [hunk 4]
```diff
@@ -15,7 +15,7 @@ protected override List<ApprenticeshipPriceEpisodeModel> GetValidApprenticeshipP
             return dataLockValidationModel.Apprenticeship.ApprenticeshipPriceEpisodes.Where(o => !o.Removed)?.ToList();
         }
 
-        protected override bool FailedValidation(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
+        protected override bool FailedValidationAsync(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipPausedValidator.cs]: Why an Async suffix without returning a `Task`?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/BaseCourseValidator.cs [hunk 5]
```diff
@@ -38,7 +38,7 @@ private ValidationResult CreateValidationResult(DataLockValidationModel dataLock
             return result;
         }
 
-        protected virtual bool FailedValidation(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
+        protected virtual bool FailedValidationAsync(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/BaseCourseValidator.cs]: Why an Async suffix without returning a `Task`?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/MultipleLearnersValidator.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            this.calculatePeriodStartAndEndDate = calculatePeriodStartAndEndDate;
+        }
+
+        protected override DataLockErrorCode DataLockerErrorCode { get; } = DataLockErrorCode.DLOCK_08;
+        protected override List<ApprenticeshipPriceEpisodeModel> GetValidApprenticeshipPriceEpisodes(DataLockValidationModel dataLockValidationModel)
+        {
+            return dataLockValidationModel.Apprenticeship.ApprenticeshipPriceEpisodes;
+        }
+
+        protected override  bool FailedValidationAsync(DataLockValidationModel dataLockValidationModel, List<ApprenticeshipPriceEpisodeModel> validApprenticeshipPriceEpisodes)
```
Comment (carlbm) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/MultipleLearnersValidator.cs]: Why an Async suffix without returning a `Task`?

File: src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs [hunk 7]
```diff
@@ -29,11 +31,36 @@ public async Task<List<ApprenticeshipModel>> ApprenticeshipsForProvider(long ukp
...
+                 .ToListAsync()
+                 .ConfigureAwait(false);
+
+            RemoveNonActivePriceEpisodes(apprenticeships);
+
+            return apprenticeships;
+
+        }
+
+        private static void RemoveNonActivePriceEpisodes(List<ApprenticeshipModel> apprenticeships)
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs]: There is no way to filter included child data in Entity Framework hence the need for the extra method. filter them out

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/MultipleLearnersValidator.cs (hunk 6 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/MultipleLearnersValidator.cs]: Changed

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/BaseCourseValidator.cs (hunk 5 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/BaseCourseValidator.cs]: Changed

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipPausedValidator.cs (hunk 4 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipPausedValidator.cs]: Changed

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipLevyPayerValidator.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/ApprenticeshipLevyPayerValidator.cs]: Changed

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/MultipleLearnersValidatorTest.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/MultipleLearnersValidatorTest.cs]: Changed

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CalculatePeriodStartAndEndDateTest.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CalculatePeriodStartAndEndDateTest.cs]: Added

File: src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application/Repositories/ApprenticeshipRepository.cs]: I have created a separate Story to optimise this query the story number is - PV2-1082. We don't have to stop DLock 8 for it.

**Pull request merged**